code/datums/map_zones.dm

```diff
@@ -0,0 +1,634 @@
+	x_distance = high_x - low_x + 1
+	y_distance = high_y - low_y + 1
+
+/datum/virtual_level/proc/is_in_bounds(atom/Atom)
+	if(Atom.x >= low_x && Atom.x <= high_x && Atom.y >= low_y && Atom.y <= high_y && Atom.z == z_value)
+		return TRUE
+	return FALSE
+
+/datum/virtual_level/proc/get_block()
+	return block(locate(low_x,low_y,z_value), locate(high_x,high_y,z_value))
+
+/datum/virtual_level/proc/get_unreserved_block()
+	return block(locate(low_x + reserved_margin, low_y + reserved_margin, z_value), locate(high_x - reserved_margin,high_y - reserved_margin,z_value))
+
+/datum/virtual_level/proc/get_center()
+	return locate(round((low_x + high_x) / 2), round((low_y + high_y) / 2), z_value)
+
+/datum/virtual_level/proc/get_random_position()
+	return locate(rand(low_x, high_x), rand(low_y, high_y), z_value)
+
+/datum/virtual_level/proc/get_below_turf(turf/Turf)
+	if(!down_linkage)
+		return
+	var/abs_x = Turf.x - low_x
+	var/abs_y = Turf.y - low_y
+	return locate(down_linkage.low_x + abs_x, down_linkage.low_y + abs_y, down_linkage.z_value)
+
+/datum/virtual_level/proc/get_above_turf(turf/Turf)
+	if(!up_linkage)
+		return
+	var/abs_x = Turf.x - low_x
+	var/abs_y = Turf.y - low_y
+	return locate(up_linkage.low_x + abs_x, up_linkage.low_y + abs_y, up_linkage.z_value)
+
+/datum/virtual_level/proc/get_client_mobs()
+	return get_alive_client_mobs() + get_dead_client_mobs()
+
+/datum/virtual_level/proc/get_alive_client_mobs()
+	. = list()
+	for(var/mob/Mob as anything in SSmobs.clients_by_zlevel[z_value])
+		if(is_in_bounds(Mob))
+			. += Mob
+
+/datum/virtual_level/proc/get_dead_client_mobs()
+	. = list()
+	for(var/mob/Mob as anything in SSmobs.dead_players_by_zlevel[z_value])
+		if(is_in_bounds(Mob))
+			. += Mob
+
+/datum/virtual_level/proc/get_mind_mobs()
+	. = list()
+	for(var/mob/living/living_mob as anything in GLOB.mob_living_list)
+		if(!living_mob.mind)
+			continue
+		if(is_in_bounds(living_mob))
+			. += living_mob
+
+/datum/virtual_level/proc/fill_in(turf/turf_type, area/area_override)
+	var/area/area_to_use = null
+	if(area_override)
+		if(ispath(area_override))
+			area_to_use = new area_override
+		else
+			area_to_use = area_override
+
+	if(area_to_use)
+		for(var/turf/iterated_turf as anything in get_block())
+			var/area/old_area = get_area(iterated_turf)
+			area_to_use.contents += iterated_turf
+			iterated_turf.change_area(old_area, area_to_use)
+			CHECK_TICK
+			if(QDELETED(src))
+				return
+
+	for(var/turf/iterated_turf as anything in get_unreserved_block())
+		iterated_turf.ChangeTurf(turf_type, turf_type)
+		CHECK_TICK
+		if(QDELETED(src))
+			return
+
+/// Gets the virtual level that contains the passed atom
+/datum/controller/subsystem/mapping/proc/get_virtual_level(atom/Atom)
+	var/datum/space_level/level = z_list[Atom.z]
+	var/datum/virtual_level/sub_map
+	for(var/datum/virtual_level/iterated_zone as anything in level.virtual_levels)
+		if(iterated_zone.is_in_bounds(Atom))
+			sub_map = iterated_zone
+			break
+	return sub_map
```

Comment:
We need to optimize the feck out of this proc if possible, as it's currently the third most self-intensive proc due to sheer volume of calls.

Also there's a ton of instances where the atom passed in ends up being `null` but I don't want that checked for here as that would contradict what I just said, we need to stop the nulls at the source instead of here.

Reply:
More investigation as to why it's being called so much, and perhaps limiting the amount of virtual levels in a single physical level. There can probably be 25 transit instances at one time in one level.

I'm positive that some things calling this wouldn't even need to do that, or could use caching or other efficiency methods.

And if this will keeps being problem I know a super fast solution that comes at small memory cost.

Reply:
You might be able to offset the cost of this proc by having it call CHECK_TICK

Reply:
Nah, I think the return value is needed instantly in most cases this is called, I think making it mildly async would break things

Reply:
But also yes this kind of proc should only be called where absolutely necessary given it iterates on basically every virtual level datum, of which there can be many

Reply:
I'm positive most, if not all procs requiring this information require it immediately

Reply:
Lookup is now fast